### PR TITLE
KEP-4265: add e2e tests for ProcMountType

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

- [KEP]: https://github.com/kubernetes/enhancements/pull/4265
```

### CODE REVIEW

File: test/e2e_node/proc_mount_test.go
Comment (kannon92) [inline test/e2e_node/proc_mount_test.go:71]: This seems long for this test. 

Once the test is finished, would we wait for this pod to finish?

Comment (haircommander) [reply test/e2e_node/proc_mount_test.go:71]: no the pods will be killed when the framework is torn down. It is a bit high, I copied it from a checkpoint test and it doesn't seem to negatively effect that one

Comment (haircommander) [reply test/e2e_node/proc_mount_test.go:71]: (also it passes locally and doesn't wait 10000 seconds :) )

**Comment by haircommander** [comment]
/retest
[[sig-storage] HostPath should support r/w [NodeConformance]](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123303/pull-kubernetes-node-e2e-containerd/1758537376822464512)

File: test/e2e/windows/security_context.go
Comment (kannon92) [inline test/e2e/windows/security_context.go:135]: cc @marosset 

how can we verify that these tests are working?

**Comment by kannon92** [comment]
/cancel lgtm

/hold

@haircommander these tests seem to say something is wrong with the PR

**Comment by haircommander** [comment]
ah I think we need https://github.com/kubernetes/test-infra/pull/32043

**Comment by haircommander** [comment]
/test pull-kubernetes-e2e-capz-windows-master

the proc mount tests passed in the alpha test :)

**Comment by kannon92** [comment]
/lgtm
/assign @mrunalp 

for sig-node side.

I think we need a sig-windows person to review/approve the windows test.

File: test/e2e_node/proc_mount_test.go
Comment (SergeyKanzhelev) [inline test/e2e_node/proc_mount_test.go:46]: let's use new way defining feature - there is a file listing all the features as constants

Comment (haircommander) [reply test/e2e_node/proc_mount_test.go]: done!

**Comment by haircommander** [comment]
I'm carrying https://github.com/kubernetes/kubernetes/pull/123180 here (exactly, not fixing the conflict with master) because we can use the node status object with runtime class info to see whether a node supports userns. 

That, in conjunction with https://github.com/kubernetes/test-infra/pull/32153 will allow us to skip the tests on nodes it's not supported, and use it in places it is.

**Comment by haircommander** [comment]
finally! proc mount tests are passing :) relies on https://github.com/kubernetes/kubernetes/pull/123180
/hold

**Comment by apevec** [comment]
> finally! proc mount tests are passing :) relies on #123180 /hold

123180 is merged, can hold be removed?

**Comment by haircommander** [comment]
also need https://github.com/kubernetes/test-infra/pull/32695

**Comment by haircommander** [comment]
```
  Timeline >>
  I0703 17:12:23.388143    3052 e2e_node_suite_test.go:283] Stopping node services...
  I0703 17:12:23.388158    3052 server.go:260] Kill server "services"
  I0703 17:12:23.388178    3052 server.go:297] Killing process 4094 (services) with -TERM
  E0703 17:12:23.462600    3052 services.go:93] Failed to stop services: error stopping "services": waitid: no child processes
  I0703 17:12:23.462614    3052 server.go:260] Kill server "kubelet"
  I0703 17:12:23.480003    3052 server.go:321] Stopping systemd unit for server "kubelet" with unit name: "kubelet-20240703T171013.service"
  I0703 17:12:23.567660    3052 services.go:153] Fetching log files...
  I0703 17:12:23.568186    3052 services.go:162] Get log file "containerd.log" with journalctl command [-u containerd].
  I0703 17:12:23.578534    3052 services.go:162] Get log file "containerd-installation.log" with journalctl command [-u containerd-installation].
  I0703 17:12:23.585693    3052 services.go:162] Get log file "crio.log" with journalctl command [-u crio].
  I0703 17:12:23.691866    3052 services.go:162] Get log file "kern.log" with journalctl command [-k].
  I0703 17:12:23.723131    3052 services.go:162] Get log file "cloud-init.log" with journalctl command [-u cloud*].
  E0703 17:12:23.733130    3052 services.go:165] failed to get "cloud-init.log" from journald: Failed to add filter for units: No data available
  , exit status 1
  I0703 17:12:23.733158    3052 e2e_node_suite_test.go:288] Tests Finished
  << Timeline
------------------------------
[ReportAfterSuite] PASSED [0.019 seconds]
[ReportAfterSuite] Kubernetes e2e JUnit report
k8s.io/kubernetes/test/e2e/framework/test_context.go:612
------------------------------

Ran 2 of 590 Specs in 114.739 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 588 Skipped
```
nice!

/retest
/test pull-kubernetes-node-e2e-containerd-alpha-features 
/test pull-kubernetes-node-kubelet-serial-containerd-alpha-features

**Comment by haircommander** [comment]
[failures](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123303/pull-kubernetes-node-e2e-containerd-alpha-features/1809220247182381056) don't seem related

/skip

LGTM, since we worked on this PR together I can't give the tag haha

File: pkg/apis/core/types.go
Comment (saschagrunert) [inline pkg/apis/core/types.go:4907]: Getting API approval is hard. Do you think it would be better to carve-out those changes into a separate PR?

Comment (sohankunkerkar) [reply pkg/apis/core/types.go:4907]: https://github.com/kubernetes/kubernetes/pull/126034

**Comment by kannon92** [comment]
I’m out of the loop on these changes. Could you update your description to reflect what this PR is. The change request is much larger than e2e tests and it isn’t clear if you want the API PR first and then you will rebase on that or you want this PR reviewed as is.

**Comment by haircommander** [comment]
yeah this is carrying https://github.com/kubernetes/kubernetes/pull/126034 so it has API changes. I think we should merge that one first and then this after

**Comment by sohankunkerkar** [comment]
> I’m out of the loop on these changes. Could you update your description to reflect what this PR is. The change request is much larger than e2e tests and it isn’t clear if you want the API PR first and then you will rebase on that or you want this PR reviewed as is.

I can't update the description, but as @haircommander mentioned, we need to get this [PR](https://github.com/kubernetes/kubernetes/pull/126034) merged before proceeding with this one. I will drop the API change commit once that PR is merged.

File: test/e2e_node/proc_mount_test.go
Comment (mrunalp) [inline test/e2e_node/proc_mount_test.go:50]: nit: drop empty line

**Review by mrunalp — APPROVED** [review]

**Comment by mrunalp** [comment]
@dims ptal

File: test/e2e_node/proc_mount_test.go
Comment (sohankunkerkar) [reply test/e2e_node/proc_mount_test.go]: Fixed

**Comment by dims** [comment]
/approve
/hold

(for tests to go green and for @mrunalp's `/lgtm`)

**Pull request merged**
